blobnet/tests/cached_test.rs

```diff
@@ -70,3 +74,73 @@ async fn concurrent_cacheable_reads() -> Result<()> {
     assert_eq!(total_net_out_bytes, s1.len());
     Ok(())
 }
+
+#[tokio::test]
+async fn request_cancellation() -> Result<()> {
+    // Create a mock provider that captures requests
+    let mock_provider = Arc::new(MockProvider::default());
+
+    // Create a cached provider wrapping the mocker provider
+    let dir = tempdir().unwrap();
+    let cached_provider = Arc::new(provider::Cached::new(
+        mock_provider.clone(),
+        dir.path().join("cache"),
+        1 << 21,
+    ));
+
+    // Create a dummy blob
+    let data = "hello world!";
+    let hash = format!("{:x}", Sha256::new().chain_update(data).finalize());
+    let range = (0, data.len() as u64);
+
+    let client = cached_provider.clone();
+
+    // Submit the first request
+    let mut f1 = {
+        let client = client.clone();
+        let hash = hash.clone();
+        task::spawn(async move { client.get(&hash, Some(range)).await })
+    };
+
+    // Get the resulting inner request
+    let inner_request = match mock_provider.requests.recv().await? {
+        Request::Get(_, response) => response,
+        _ => bail!("Unexpected inner request type!"),
+    };
+
+    // Submit another request
+    let mut f2 = {
+        let client = client.clone();
+        let hash = hash.clone();
+        task::spawn(async move { client.get(&hash, Some(range)).await })
+    };
+
+    // Check that we did not get another inner request and that neither request
+    // completes
+    let sleep = time::sleep(Duration::from_millis(200));
+    tokio::select! {
+        _ = &mut f1 => {
+            bail!("Unexpected f1 completion!");
+        },
+        _ = &mut f2 => {
+            bail!("Unexpected f2 completion!");
+        },
+        _ = mock_provider.requests.recv() => {
+            bail!("Unexpected f2 completion!");
```

Comment:
```suggestion
            bail!("Unexpected additional request to inner provider!");
```

If I understand things correctly, the above description describes the error because this `.recv()` isn't checking for f2 completion?